src/DeviceInterface.cpp

```diff
@@ -0,0 +1,133 @@
+#include "JITModule.h"
+#include "StmtCompiler.h"
+#include "Target.h"
+
+#include "runtime/HalideRuntime.h"
+#include "runtime/HalideRuntimeCuda.h"
+#include "runtime/HalideRuntimeOpenCL.h"
+#include "runtime/HalideRuntimeOpenGL.h"
+
+using namespace Halide;
+using namespace Halide::Internal;
+
+namespace {
+
+template <typename fn_type>
+bool lookup_runtime_routine(const char *name, const Target &target, fn_type &result) {
+    StmtCompiler stmt_compiler(target);
+    std::vector<JITModule> runtime(JITSharedRuntime::get(stmt_compiler.get_codegen(), target));
```

Comment:
> Note also that the routines to get an interface (e.g. halide_cuda_device_interface) can legitimately be called before any runtime has been loaded for other reasons to get a device interface to use for halide_device_malloc.

Yeah, the only routines I allowed to not create a runtime were halide_device_release and halide_device_free. Also, in my branch, I returned an error from halide_device_release if buffer_t::dev is non-zero, but the runtime was not created already. That's not really a fix... but its something I guess.

I don't really have a good answer to your other point. Maybe we should just check if the dev field is zero and skip free, otherwise just always create the runtime.